src/System.Private.CoreLib/shared/System/Text/ASCIIEncoding.cs

```diff
@@ -193,22 +277,96 @@ public override unsafe int GetByteCount(ReadOnlySpan<char> chars)
+                    argument: (chars is null) ? ExceptionArgument.chars : ExceptionArgument.bytes,
+                    resource: ExceptionResource.ArgumentNull_Array);
+            }
+
+            if ((charCount | byteCount) < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(
+                    argument: (charCount < 0) ? ExceptionArgument.charCount : ExceptionArgument.byteCount,
+                    resource: ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
+            }
 
-            return GetBytes(chars, charCount, bytes, byteCount, null);
+            return GetBytesCommon(chars, charCount, bytes, byteCount);
         }
 
         public override unsafe int GetBytes(ReadOnlySpan<char> chars, Span<byte> bytes)
         {
-            fixed (char* charsPtr = &MemoryMarshal.GetNonNullPinnableReference(chars))
-            fixed (byte* bytesPtr = &MemoryMarshal.GetNonNullPinnableReference(bytes))
+            // It's ok for us to operate on null / empty spans.
+
+            fixed (char* charsPtr = &MemoryMarshal.GetReference(chars))
+            fixed (byte* bytesPtr = &MemoryMarshal.GetReference(bytes))
+            {
+                return GetBytesCommon(charsPtr, chars.Length, bytesPtr, bytes.Length);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe int GetBytesCommon(char* pChars, int charCount, byte* pBytes, int byteCount)
+        {
+            // Common helper method for all non-EncoderNLS entry points to GetBytes.
+            // A modification of this method should be copied in to each of the supported encodings: ASCII, UTF8, UTF16, UTF32.
+
+            Debug.Assert(charCount >= 0, "Caller should't specify negative length buffer.");
+            Debug.Assert(pChars != null || charCount == 0, "Input pointer shouldn't be null if non-zero length specified.");
+            Debug.Assert(byteCount >= 0, "Caller should't specify negative length buffer.");
+            Debug.Assert(pBytes != null || byteCount == 0, "Input pointer shouldn't be null if non-zero length specified.");
+
+            // First call into the fast path.
+
+            int numElementsConverted = (int)ASCIIUtility.NarrowUtf16ToAscii(pChars, pBytes, Math.Min((uint)charCount, (uint)byteCount));
+
+            if (numElementsConverted == charCount)
+            {
+                // All elements converted - return immediately.
+
+                return numElementsConverted;
+            }
+            else
+            {
+                // Simple narrowing conversion couldn't operate on entire buffer - invoke fallback.
+
+                return GetBytesWithFallback(pChars, charCount, pBytes, byteCount, numElementsConverted, numElementsConverted);
+            }
+        }
+
+        private protected sealed override unsafe int GetBytesNoFallbackBuffer(char* pChars, int charCount, byte* pBytes, int byteCount, EncoderFallback fallback, out int charsConsumed)
```

Comment:
I think `No` in the method name is hard to understand, and rare/non-existent - I was not able to find a similar name with `No` in the middle anywhere in CoreFX.

The important thing about this method is that it is the fast path. Would it make sense to call it `GetBytesOptimized` ?

Naming is hard...

Reply:
I have an idea for how this can be simplified and will bring it through the next iteration.